### PR TITLE
Fix exporters to use fully qualified filenames

### PR DESCRIPTION
**Pull request by Teknikaali**
This PR fixes exporters to use fully qualified filenames instead of simple type names.
## Issue
#529

##  Features
- Changes how `ExporterBase` handles file naming:
  - Uses `Type.FullName` instead of `Type.Name` as part of the filename
  - Uses `Summary.Title` as the filename, if multiple benchmarks have been joined into a single `Summary`
    - This convention follows the way `Summary.Title` is generated in [`BenchmarkRunnerCore r60`](https://github.com/dotnet/BenchmarkDotNet/blob/174c19dced6883c9f5d810303b7e969739e1cc2b/src/BenchmarkDotNet.Core/Running/BenchmarkRunnerCore.cs#L60)
- Adds two integration tests for the aforementioned features inside `BenchmarkDotNet.IntegrationTests.ExporterIOTests`:
  1) Asserts the exporter uses fully qualified type name as part of the filename
  2) Asserts joined summaries use `Summary.Title` as filename instead of the fully qualified type name

### Example filename
Before: `IntroExporters-report-brief.json`
After: `BenchmarkDotNet.Samples.Intro.IntroExporters-report-brief.json`

## Other notes
- These changes won't take into account multiple classes with a same name inside different namespaces when _joining summaries of multiple benchmarks_
  - For example benchmarks inside `NamespaceA.ClassA` and `NamespaceB.ClassA` wouldn't be considered as two different benchmark classes _for the filename_ when joining. `ClassA-report.txt` would be used as the filename instead of `BenchmarkRun-001-2017-09-21-03-42-28-report.txt`
> I think this is a rare corner case so I left its handling out of for now

I think that's about it! ✔️
Let me know if I missed something or if something needs improvement.

### CODE REVIEW

File: tests/BenchmarkDotNet.IntegrationTests/ExporterIOTests.cs [hunk 1]
```diff
@@ -90,5 +91,47 @@ public override void ExportToLog(Summary summary, ILogger logger)
                 ExportCount++;
             }
         }
+
+        [Fact]
+        public void ExporterUsesFullyQualifiedTypeNameAsFileName()
+        {
+            var exporter = new MockExporter();
+            var summary = CanExecute<ClassA>();
```
Comment (adamsitnik) [inline tests/BenchmarkDotNet.IntegrationTests/ExporterIOTests.cs]: could you create some `Summary` without running the benchmarks?

We are currently having a problem with too long builds. We have too many integration tests that run benchmarks and we exceed the 1h we have in our current plan :/

Comment (Teknikaali) [reply tests/BenchmarkDotNet.IntegrationTests/ExporterIOTests.cs]: Yes I'll look into it! AppVeyor currently reports these durations for the added tests:

Test | Duration
----- | ---------
ExporterUsesFullyQualifiedTypeNameAsFileName | 28 sec 293 ms
ExporterUsesSummaryTitleAsFileNameWhenBenchmarksJoinedToSingleSummary | 3 min 51 sec 219 ms

I suppose its possible to get these down to a couple of milliseconds each, if I mock out the `Summary` correctly.

Comment (adamsitnik) [reply tests/BenchmarkDotNet.IntegrationTests/ExporterIOTests.cs]: > I suppose its possible to get these down to a couple of milliseconds each, if I mock out the Summary correctly.

exactly! thanks for understanding!

**Comment by adamsitnik** [comment]
@Teknikaali thanks a lot!
